src/db/mapping.cr

```diff
@@ -0,0 +1,136 @@
+module DB
+
+  # The `DB.mapping` macro defines how an object is built from a DB::ResultSet.
+  #
+  # It takes hash literal as argument, in which attributes and types are defined.
+  # Once defined, `DB::ResultSet#read(t)` populates properties of the class from the
+  # result set.
+  #
+  # ```crystal
+  # require "db"
+  #
+  # class Employee
+  #   DB.mapping({
+  #     title: String,
+  #     name:  String,
+  #   })
+  # end
+  #
+  # employees = Employee.from_rs(db.query("SELECT title, name FROM employees"))
+  # employees[0].title # => "Manager"
+  # employees[0].name  # => "John"
+  # ```
+  #
+  # Attributes not mapped with `DB.mapping` are not defined as properties.
+  # Also, missing attributes raise a `DB::Exception`.
+  #
+  # You can also define attributes for each property.
+  #
+  # ```crystal
+  # class Employee
+  #   DB.mapping({
+  #     title: String,
+  #     name:  {
+  #       type:    String,
+  #       nilable: true,
+  #       key:     "firstname",
+  #     },
+  #   })
+  # end
+  # ```
+  #
+  # Available attributes:
+  #
+  # * *type* (required) defines its type. In the example above, *title: String* is a shortcut to *title: {type: String}*.
+  # * *nilable* defines if a property can be a `Nil`.
+  # * **default**: value to use if the property is missing in the result set, or if it's `null` and `nilable` was not set to `true`. If the default value creates a new instance of an object (for example `[1, 2, 3]` or `SomeObject.new`), a different instance will be used each time a row is parsed.
+  # * *key* defines which column to read from a reusltset. It defaults to the name of the property.
+  # * *converter* takes an alternate type for parsing. It requires a `#from_rs` method in that class, and returns an instance of the given type.
+  #
+  # The mapping also automatically defines Crystal properties (getters and setters) for each
+  # of the keys. It doesn't define a constructor accepting those arguments, but you can provide
+  # an overload.
+  #
+  # The macro basically defines a constructor accepting a `DB::ResultSet` that reads from
+  # it and initializes this type's instance variables.
+  #
+  # This macro also declares instance variables of the types given in the mapping.
+  macro mapping(properties, strict = true)
+    {% for key, value in properties %}
+      {% properties[key] = {type: value} unless value.is_a?(HashLiteral) || value.is_a?(NamedTupleLiteral) %}
+    {% end %}
+
+    {% for key, value in properties %}
+      @{{key.id}} : {{value[:type]}} {{ (value[:nilable] ? "?" : "").id }}
+
+      def {{key.id}}=(_{{key.id}} : {{value[:type]}} {{ (value[:nilable] ? "?" : "").id }})
+        @{{key.id}} = _{{key.id}}
+      end
+
+      def {{key.id}}
+        @{{key.id}}
+      end
+    {% end %}
+
+    def self.from_rs(%rs : DB::ResultSet)
```

Comment:
this should go away, we can merge it as is, and remove it when adapting to the new query_one/query_all api.
agreed?